File: lib/channel/f_channel_subscriber.dart
import "../execution_context/f_execution_context.dart" show FExecutionContext;

///
/// Define some kind of Publish-Subscribe pattern.
///
/// See https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern
///
abstract class FSubscriberChannel<TData,
    TEvent extends FChannelSubscriberEvent<TData>> {
  void addHandler(FSubscriberChannelCallback<TData, TEvent> cb);
  void removeHandler(FSubscriberChannelCallback<TData, TEvent> cb);
}

abstract class FChannelSubscriberEvent<TData> {
  TData get data;
}

///
/// event - TEvent or FException
///
typedef FSubscriberChannelCallback<TData,
        TEvent extends FChannelSubscriberEvent<TData>>
    = Future<void> Function(
  FExecutionContext executionContext,
  dynamic eventOrException,
);

File: src/channel/FChannelSubscriber.ts
import { FException } from "../exception";
import { FExecutionContext } from "../execution_context";

/** Define some kind of Publish-Subscribe pattern. See https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern */
export interface FChannelSubscriber<
	TData,
	TEvent extends FChannelSubscriber.Event<TData> = FChannelSubscriber.Event<TData>
> {
	addHandler(cb: FChannelSubscriber.Callback<TData, TEvent>): void;
	removeHandler(cb: FChannelSubscriber.Callback<TData, TEvent>): void;
}
export namespace FChannelSubscriber {
	export interface Event<TData> {
		readonly data: TData;
	}
	export interface Callback<TData, TEvent extends Event<TData> = Event<TData>> {
		(executionContext: FExecutionContext, event: TEvent | FException): Promise<void>;
	}
}
